runtime/runtime/src/pipelining.rs

```diff
@@ -0,0 +1,321 @@
+struct PrepareTaskKey {
+    receipt_id: CryptoHash,
+    action_index: usize,
+}
+
+struct PrepareTask {
+    status: Mutex<PrepareTaskStatus>,
+    condvar: Condvar,
+}
+
+enum PrepareTaskStatus {
+    Pending,
+    Working,
+    Prepared(Box<dyn PreparedContract>),
+    Finished,
+}
+
+impl ReceiptPreparationPipeline {
+    pub(crate) fn new(
+        config: Arc<RuntimeConfig>,
+        contract_cache: Option<Box<dyn ContractRuntimeCache>>,
+        chain_id: String,
+        protocol_version: u32,
+        storage: near_store::contract::Storage,
+    ) -> Self {
+        Self {
+            map: Default::default(),
+            block_accounts: Default::default(),
+            config,
+            contract_cache,
+            chain_id,
+            protocol_version,
+            storage,
+        }
+    }
+
+    /// Submit a receipt to the "pipeline" for preparation of likely eventual execution.
+    ///
+    /// Note that not all receipts submitted here must be actually handled in some way. That said,
+    /// while it is perfectly fine to not use the results of submitted work (e.g. because a
+    /// applying a chunk ran out of gas or compute cost,) this work would eventually get lost, so
+    /// for the most part it is best to submit work with limited look-ahead.
+    ///
+    /// Returns `true` if the receipt is interesting and work has been scheduled for its
+    /// preparation.
+    pub(crate) fn submit(
+        &mut self,
+        receipt: &Receipt,
+        account: &Account,
+        view_config: Option<ViewConfig>,
+    ) -> bool {
+        let account_id = receipt.receiver_id();
+        if self.block_accounts.contains(account_id) {
+            return false;
+        }
+        let actions = match receipt.receipt() {
+            ReceiptEnum::Action(a) | ReceiptEnum::PromiseYield(a) => &a.actions,
+            ReceiptEnum::Data(_) | ReceiptEnum::PromiseResume(_) => return false,
+        };
+        let mut any_function_calls = false;
+        for (action_index, action) in actions.iter().enumerate() {
+            let account_id = account_id.clone();
+            match action {
+                Action::DeployContract(_) => {
+                    // FIXME: instead of blocking these accounts, move the handling of
+                    // deploy action into here, so that the necessary data dependencies can be
+                    // established.
+                    self.block_accounts.insert(account_id);
+                    break;
+                }
+                Action::FunctionCall(function_call) => {
+                    let key = PrepareTaskKey { receipt_id: receipt.get_hash(), action_index };
+                    let gas_counter = self.gas_counter(view_config.as_ref(), function_call.gas);
+                    let entry = match self.map.entry(key) {
+                        std::collections::btree_map::Entry::Vacant(v) => v,
+                        // Already been submitted.
+                        // TODO: Warning?
```

Comment:
I guess we don't want warning here since having multiple function calls in the same receipt is normal and it would trigger that case, right?

Reply:
The key used in the map is receipt hash + action index, so this case being hit would mean that the caller is submitting the same receipt (with all its actions) multiple times in the same chunk, possibly with the intent to execute the actions multiple times. This won't work in the current implementation as the prepared contracts cannot be reused.